mindsdb/api/executor/sql_query/steps/fetch_dataframe_partition.py

```diff
@@ -87,6 +95,56 @@ def call(self, step: FetchDataframeStepPartition) -> ResultSet:
         else:
             return self.fetch_iterate(run_query, query, on_error=on_error)
 
+    def repeat_till_reach_limit(self, step, limit):
+        first_table_limit = limit * 2
+        dn = self.session.datahub.get(step.integration)
+
+        query = step.query
+
+        # fill params
+        query, context_callback = query_context_controller.handle_db_context_vars(query, dn, self.session)
+
+        try_num = 1
+        while True:
+            self.substeps = copy.deepcopy(step.steps)
+            query2 = copy.deepcopy(query)
+
+            if first_table_limit is not None:
+                query2.limit = Constant(int(first_table_limit))
+            else:
+                query2.limit = None
+
+            response = dn.query(query=query2, session=self.session)
+            df = response.data_frame
+
+            result = self.exec_sub_steps(df)
+
+            if len(result) >= limit or len(df) < first_table_limit or first_table_limit is None:
+                # we have enough results
+                #  OR first table doesn't return requested count of rows
+                #  OR it is a flag to stop
+                result = result[:limit]
+                break
+
+            if try_num > 3:
+                # the last try without the limit
+                first_table_limit = None
+                continue
+
+            # no enough results
+            if len(result) > 0:
+                # forecast the required limit (depending on how much row we don't have)
+                first_table_limit = first_table_limit * limit / len(result) * try_num + 10**try_num
```

Comment:
first_table_limit must be rounded here, instead of line 113
Otherwise check in line 122 may be wrong (for example `len([1,2,3]) < 3.1` )